enhancements/etcd/automated-cluster-backups.md

```diff
@@ -0,0 +1,137 @@
+
+## Goals
+
+1.  Automatic periodic backups and automatic pruning
+1.  Make most recent backups readily available
+1.  Eliminate SSH requirement
+
+## Non-Goals
+
+*  Automated restoration is not considered for this enhancement.
+
+## Currently Supported Functionality 
+
+Currently the scripts available support the following functionality:
+### Cluster Backup 
+Takes a snapshot of cluster’s etcd data along with static-pod-resources at the time of the backup. 
+
+### Cluster Restore 
+Restores the etcd data from a backup snapshot. It also restores the static pod resources while deleting all the newer
+revisions.
+
+## Proposal
+
+### A new container is added to etcd static pod to run automated backups. 
+The objectives of the new container are:
+1.  On all etcd static pods, run a subcommand of cluster-etcd-operator to produce automated backups.
+1. On all masters it creates a new backup revision `/etc/kubernetes/static-pod-backups/backup-N`.
+1. On all masters it write `/etc/kubernetes/static-pod-backups/backup-N/backup.env` file containing 4 environmental
+   variables `CREATED`, `OCP_VERSION`, `ETCD_REVISION` and `APISERVER_REVISION`.
+1. On all masters take an etcd snapshot `etcdctl snapshot save
+    /etc/kubernetes/static-pod-backups/backup-N/etcd-data/backup.db`.
+1. On all masters copy all static pod resources  to `/etc/kubernetes/static-pod-backups/backup-N/`.
+1. On all masters symbolically link the directory `/etc/kubernetes/static-pod-backups/latest-backup` with the directory
+   containing the most recent `/etc/kubernetes/static-pod-manifests/backup-N` directory.
+1.  Also be responsible for pruning older backups to keep no more than X number of backups.
+
+###
+
+## User Stories [optional]
+
+### Security
+
+Your clusters backup data is as secure as your cluster. If someone were to root the system they would have direct
+access to all data.
+
+### Availability
+
+Your data is as resilient as your cluster. We make N copies of your data so in the case of failure you dont have to
+worry about your last backup location.
+
+### Recovery Automation
+
+If the cluster were to lose quorum and every master is seeded with data required to restore. Automation of recovery
+tasks becomes easier.
+
+
+## Implementation Plan
+
+1. Create a subcommand to cluster-etcd-operator to run as a container in the etcd static pod.
+2. It runs in an auto-pilot mode without requiring adding any new OpenShift API (for 4.4)
+3. In future releases, cluster-etcd-operator will subsume this functionality with a backup controller (4.6+).
+
+## Implementation Details/Notes/Constraints
+
+*  Add a subcommand to `cluster-etcd-operator`
+*  Change the etcd spec `etcd-pod.yaml` to run the subcommand as a separate container.
+
+## Risks and Mitigations
+
+## Design Details
+
+## Test Plan
+
+## Graduation Criteria
+
+## Upgrade / Downgrade Strategy
+
+## Version Skew Strategy
+
+
+## Implementation History
+
+## Drawbacks
+
+## Alternatives
```

Comment:
Need to talk about the non-sidecar approaches here

Reply:
Listed them. Will elaborate.